util/src/main/java/io/grpc/util/DeterministicSubsettingLoadBalancer.java

```diff
@@ -0,0 +1,210 @@
+
+    ResolvedAddresses subsetAddresses = buildSubsets(resolvedAddresses, config);
+
+    switchLb.handleResolvedAddresses(
+        subsetAddresses.toBuilder()
+            .setLoadBalancingPolicyConfig(config.childPolicy.getConfig())
+            .build());
+    return true;
+  }
+
+  // implements the subsetting algorithm, as described in A68:
+  // https://github.com/grpc/proposal/pull/383
+  private ResolvedAddresses buildSubsets(
+      ResolvedAddresses allAddresses, DeterministicSubsettingLoadBalancerConfig config) {
+    // The map should only retain entries for addresses in this latest update.
+    ArrayList<SocketAddress> addresses = new ArrayList<>();
+    for (EquivalentAddressGroup addressGroup : allAddresses.getAddresses()) {
+      addresses.addAll(addressGroup.getAddresses());
+    }
+
+    if (addresses.size() <= config.subsetSize) {
+      return allAddresses;
+    }
+    if (config.sortAddresses) {
+      // If we sort, we do so via the string representation of the SocketAddress.
+      addresses.sort(new AddressComparator());
+    }
+
+    Integer backendCount = addresses.size();
+    Integer subsetCount = backendCount / config.subsetSize;
+
+    Integer round = config.clientIndex / subsetCount;
+
+    Integer excludedCount = backendCount % config.subsetSize;
+    Integer excludedStart = (round * excludedCount) % backendCount;
+    Integer excludedEnd = (excludedStart + excludedCount) % backendCount;
+    if (excludedStart <= excludedEnd) {
+      List<SocketAddress> subList = addresses.subList(0, excludedStart);
+      subList.addAll(addresses.subList(excludedEnd, backendCount));
+      addresses = new ArrayList<>(subList);
+    } else {
+      addresses = new ArrayList<>(addresses.subList(excludedEnd, excludedStart));
+    }
+
+    Random r = new Random(round);
+    Collections.shuffle(addresses, r);
+
+    Integer subsetId = config.clientIndex % subsetCount;
+
+    Integer start = subsetId * config.subsetSize;
+    Integer end = start + config.subsetSize;
+
+    List<SocketAddress> subset = addresses.subList(start, end);
+
+    ArrayList<EquivalentAddressGroup> eaglist = new ArrayList<>();
+
+    // Create new EAGs per address
+    for (SocketAddress addr : subset) {
+      eaglist.add(new EquivalentAddressGroup(addr));
+    }
+
+    ResolvedAddresses.Builder builder = allAddresses.toBuilder();
+    return builder.setAddresses(eaglist).build();
+  }
+
+  @Override
+  public void handleNameResolutionError(Status error) {
+    switchLb.handleNameResolutionError(error);
+  }
+
+  @Override
+  public void shutdown() {
+    switchLb.shutdown();
+  }
+
+  public DeterministicSubsettingLoadBalancer(Helper helper) {
+    switchLb = new GracefulSwitchLoadBalancer(checkNotNull(helper, "helper"));
+  }
+
+  @VisibleForTesting
+  static class AddressComparator implements Comparator<SocketAddress> {
+    // For consistency with the golang instrumentation, this assumes toString is overridden such
```

Comment:
This isn't consistent with golang. I know this will change to hashing, but there will still be some similar issues. Maybe we'll be forced to use `address.hashCode()`, or maybe it doesn't matter because we just have to be consistent within a single implementation. See https://github.com/grpc/grpc-java/commit/f866c805c2f78271de9f2b61254363d009cee8c6 where I mention some issues with the string representation.